src/utils/chartJs/CustomTooltip.jsx

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import { withStyles } from '@material-ui/core/styles';
+
+const styles = {
+  tooltip: {
+    position: 'absolute',
+    padding: '6px',
+    background: 'rgba(0, 0, 0, .8)',
+    color: 'white',
+    borderRadius: '4px',
+    pointerEvents: 'none',
+  },
+  tooltipKey: {
+    display: 'inline-block',
+    width: '10px',
+    height: '10px',
+    marginRight: '10px',
+  },
+};
+
+class CustomTooltip extends React.Component {
+  constructor(props) {
+    super(props);
+    this.tooltip = React.createRef();
+    this.state = { width: 0, height: 0 };
+  }
+
+  componentDidMount() {
+    this.setState({
+      width: this.tooltip.current.clientWidth,
+      height: this.tooltip.current.clientHeight,
+    });
+  }
+
+  render() {
+    const { classes, tooltipModel, series, canvas } = this.props;
+
+    if (tooltipModel.opacity === 0) return null;
+
+    let top = canvas.offsetTop + tooltipModel.caretY;
+    let left = canvas.offsetLeft + tooltipModel.caretX;
+
+    if (tooltipModel.yAlign === 'bottom') {
+      top -= this.state.height;
+    } else if (tooltipModel.yAlign === 'center') {
+      top -= 0.5 * this.state.height;
+    }
+
+    if (tooltipModel.xAlign === 'right') {
```

Comment:
`if/else` blocks should always trigger the question: "why?"

The words "left", "center", and "right" refer to relative positions (0.0, 0.5, 1.0) respectively.  That is not only true here, 
but everywhere we use CSS.  With `CSS = {left: 0, center:0.5, right: 1}`, we can write one line:

    left -= CSS[tooltipModel.xAlign]*this.state.width;

Only make this change if you want.

Reply:
Nice one, I love how there's always something to improve :smiley: 
I will change only the name ('CSS'), because we also need  another for 'top', 'center', 'bottom'. But before that what about using this:
```
let top = canvas.offsetTop + tooltipModel.caretY;
top -=
      { top: 0, center: 0.5, bottom: 1 }[tooltipModel.yAlign] *
      this.state.height;
```
or even more compact:
```
const top =
      canvas.offsetLeft + tooltipModel.caretX -
      { top: 0, center: 0.5, bottom: 1 }[tooltipModel.yAlign] *
        this.state.height;
```
would be alright? or is it too messy?

Reply:
yes, either is fine.  I like that you inlined the "CSS" structure.